### PR TITLE
python: Enhance syntax highlighting for type hints

### PR DESCRIPTION
**Pull request by ClanEver**
Release Notes:

- Python: Improved syntax highlighting for type hints.

# Before
image

# After
image

---
Why manual recursion?
- Due to tree-sitter grammar not supporting recursion in query (https://github.com/tree-sitter-grammars/tree-sitter-lua/issues/24), currently only manual recursion is possible (refer to https://github.com/projekt0n/github-nvim-theme/pull/250/files).

<br/>

Unable to highlight when simple structures appear before complex structures, example:
```python
def t() -> str | dict[int, dict[int, dict[int, str]]]:
    pass
```
Because complex structures are parsed as `subscript` rather than `generic_type` by tree-sitter in this case ☹

<br/>

Related:

- https://github.com/zed-industries/zed/issues/14715
- [Union Type (Python Doc)](https://docs.python.org/3/library/stdtypes.html#union-type)
- [Type parameter lists (Python Doc)](https://docs.python.org/3/reference/compound_stmts.html#type-parameter-lists)

### CODE REVIEW

File: crates/languages/src/python/highlights.scm [hunk 1]
```diff
@@ -1,6 +1,59 @@
 (parameter (identifier) @variable)
 (attribute attribute: (identifier) @property)
 (type (identifier) @type)
+(generic_type (identifier) @type)
+
+; Union type X | Y (up to 8 types)
+(type
```
Comment (maxdeviant) [inline crates/languages/src/python/highlights.scm:7]: @maxbrunsfeld Do you have thoughts on this approach?

Comment (maxbrunsfeld) [reply crates/languages/src/python/highlights.scm:7]: I don't think we should land this code. I get that it's working around a limitation in the query system right now, but I think we just need to fix that limitation.

I worry this is going to have a negative performance impact, and it's also just hard to understand and maintain.

**Comment by maxbrunsfeld** [comment]
@ClanEver Can you back out the one complex pattern from the query (sorry that it's not possible to highlight union types cleanly with our current scheme) so that we can land the other changes in the query?

**Comment by ClanEver** [comment]
@maxbrunsfeld Would it be acceptable to reduce the depth to 3?
```Scheme
; Union type X | Y (up to 3 types)
(type
  (binary_operator
    left: [
      (binary_operator
        left:  (_) @type
        right: (_) @type
      ) @type
      (_) @type
    ]
    right: (_) @type
  ) @type
)
```

**Review by maxdeviant — APPROVED** [review]
Thank you!

I removed the union type highlighting until we can come up with a better solution.

**Pull request merged**